enhancements/storage/csi-driver-operator-merge.md

```diff
@@ -0,0 +1,530 @@
+       from `openshift/aws-ebs-csi-driver-operator` jobs and start promoting it from `openshift/csi-operator`.
+    2. PR against openshift/ocp-build-data to switch the source github repo + Dockerfile
+       for `ose-aws-ebs-csi-driver-operator-container` image.
+    3. Somehow coordinated merge of these two.
+      * _Can we merge 1. before 2. to see if / how it breaks CI builds? We could be able to revert back to the working
+        config without any extra approvals._
+
+* After the switch, we start actually refactoring and merging the operator code to shared packages and so on. At this
+  time, we will have CI in place for all our PRs in the repo. We will re-use `Dockerfile.aws-ebs`
+  and `Dockerfile.aws-ebs.test` to build the image, so we don't need to change anything in ART build data.
+
+  After the refactoring, the repo should look like this:
+  ```
+  ├── assets
+  │   ├── assets.go
+  │   └── generated   # We want to generate the assets, see below.
+  │       └── aws-ebs
+  │           ├── cabundle_cm.yaml
+  │           └── ... # other AWS EBS driver assets
+  ├── cmd
+  │   └── aws-ebs-csi-driver-operator
+  │       └── main.go
+  ├── pkg
+  │   ├── aws-ebs
+  │   │   └── # AWS EBS specific-code
+  │   └── common
+  │       └── starter.go # and any other code shared by all CSI drivers
+  ├── test
+  │   └── e2e
+  │       └── aws-ebs
+  │           └── manifest.yaml
+  ├── vendor
+  ├── Dockerfile.aws-ebs
+  ├── Dockerfile.aws-ebs-test
+  ├── go.mod
+  └── go.sum
+  ```
+
+There will be two major points when things can break:
+
+1. When we switch building of `ose-aws-ebs-csi-driver-operator-container` from openshift/aws-ebs-csi-driver-operator to
+   openshift/csi-operator. Since the code will be 100% the same, we think it's safe to do. Any revert must be
+   coordinated with ART in ocp-build-data and openshift/release repositories.
+2. After our refactoring, we will switch building of `Dockerfile.aws-ebs` from `legacy/aws-ebs-csi-driver-operator/cmd/`
+   to `cmd/`. At this time we will have CI that should catch any breakage. Any revert must be done in
+   openshift/csi-operator repo, which is under our control.
+
+All CSI driver operators will be merged in a similar way, i.e. in `legacy/` directory first.
+
+Drawbacks:
+
+* When anything goes wrong in switching the source repositories, it's ART who will need to roll back things, we cannot
+  do it ourselves.
+
+Advantages:
+
+* Keeping image names, i.e. no new Comet repos (distgits, brew packages, etc.)
+* Harder (but not impossible) to test before the switch. In ideal case, we're switching just the source repository
+  of the same image, the actual operator code should be the same.
+
+#### Gradual merge
+
+We will merge the operators one by one, starting with AWS EBS. Second will be Azure Disk and Azure File, as we need to
+support HyperShift for them and we want to share code with AWS EBS. The rest will follow as time allows, possibly over
```

Comment:
FYI, OpenStack is also looking at HyperShift (https://issues.redhat.com/browse/OSASINFRA-3184). It's still pretty much in exploration phase, so there's no immediate urgency in having the Cinder and Manila CSI operator part of the shared repo in 4.15 or 4.16, but for sure, we'll be happy once we share more code.